### ISSUE TITLE
The Lightning example doesn't train with privacy?

### ISSUE THREAD

**Issue by tholop** [pytorch/opacus#313]
## 🐛 Bug
I think that there is something strange with the PyTorch Lightning example. First, I realized that changing `max_grad_norm` had no impact on the accuracy. Then, using `noise_multiplier = 1000` gave much higher accuracy than expected (see below).

In the code, it seems that the DP version of the model is stored somewhere but never used:
https://github.com/pytorch/opacus/blob/eedb4447b4b5271591c268c6f280823b2415e134/examples/mnist_lightning.py#L107

Unless there is something subtle going on (maybe you fetch `self.dp` elsewhere in the optimizer when you detect a Ligthning module?), the forward pass is using the non-DP model instead of `self.dp["model"]`:
https://github.com/pytorch/opacus/blob/eedb4447b4b5271591c268c6f280823b2415e134/examples/mnist_lightning.py#L113

Please, did I miss anything?

## To Reproduce
Download `mnist_lightning.py` and run the following:
```
❯ python mnist_lightning.py fit --model.max_grad_norm 1 --model.noise_multiplier 1000  --trainer.max_epochs 3                           
GPU available: True, used: False                                                                                                              
TPU available: False, using: 0 TPU cores                                                                                                      
IPU available: False, using: 0 IPUs                                                                                                           
Epoch 2: 100%|████████████████████████████████████| 1500/1500 [00:27<00:00, 54.03it/s, loss=0.0488, v_num=2, train_loss=0.0685, epsilon=0.103]
Epoch 2: 100%|██████████████████████████████████████████████████████████████████████| 1500/1500 [00:25<00:00, 58.78it/s, loss=0.0377, v_num=2]
Restoring states from the checkpoint path at /home/.../lightning_logs/version_2/checkpoints/epoch=2-step=5999.ckpt                   
Loaded model weights from checkpoint at /home/.../lightning_logs/version_2/checkpoints/epoch=2-step=5999.ckpt                        
Testing:  97%|██████████████████████████████████████████████████████████████████████████████████████████▎  | 304/313 [00:01<00:00, 155.37it/s]--------------------------------------------------------------------------------
DATALOADER:0 TEST RESULTS                                                                                                                     
{'test_accuracy': 0.9854999780654907, 'test_loss': 0.04740014299750328}                                                                       
--------------------------------------------------------------------------------
```

**Comment by romovpa** [comment]
@tholop Thank you for reporting this!

Indeed, the Lightning example doesn't train with privacy. The reason is bug in DPOptimizer, fix & details: #330.

As for applying `self(data)` instead of `dp_model(data)` there is something subtle going on. The fact is both are correct and doing exactly the same. When we create `GradSampleModule` instance `dp_model`, the hooks for computing grad_samples are added to the original model. The wrapper is needed to remove those hooks. I believe it's not very intuitive and should be changed in the future.

**Comment by romovpa** [comment]
Fixed in #330
